Raise error for code that risk deadlock

albanD opened the pull request:
Fix for https://github.com/pytorch/pytorch/issues/32045

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32295 Raise error for code that risk deadlock**

Calling into the engine with the GIL can deadlock because:
- worker thread initialization acquires the GIL
- Any Node / hook can be a python function that will acquire the GIL

The choice was made here to raise an error as one of the advantage of using cpp extensions with python is to be able to release the GIL. So we prefer to educate users to do it rather than doing it under the hook.

Differential Revision: [D19430979](https://our.internmc.facebook.com/intern/diff/D19430979)

albanD committed:
Raise error for code that risk deadlock

[ghstack-poisoned]

zou3519 commented on torch/csrc/autograd/python_engine.cpp:
```diff
@@ -61,6 +61,10 @@ variable_list PythonEngine::execute(
     bool keep_graph,
     bool create_graph,
     const edge_list& outputs) {
+  TORCH_CHECK(!PyGILState_Check(), "The autograd engine is called while holding the GIL. If you are using the C++ "
```
nit: this reads a little better in past tense "The autograd engine was called" or "You called the autograd engine"

zou3519 reviewed (approved):
It would be nice to have a way to test this (hold the GIL while calling PythonEngine::execute and check that we're throwing an error) but that sounds difficult

albanD committed:
Update on "Raise error for code that risk deadlock"

Fix for https://github.com/pytorch/pytorch/issues/32045


Calling into the engine with the GIL can deadlock because:
- worker thread initialization acquires the GIL
- Any Node / hook can be a python function that will acquire the GIL

The choice was made here to raise an error as one of the advantage of using cpp extensions with python is to be able to release the GIL. So we prefer to educate users to do it rather than doing it under the hook.

Differential Revision: [D19430979](https://our.internmc.facebook.com/intern/diff/D19430979)

[ghstack-poisoned]

albanD committed:
Update on "Raise error for code that risk deadlock"

Fix for https://github.com/pytorch/pytorch/issues/32045


Calling into the engine with the GIL can deadlock because:
- worker thread initialization acquires the GIL
- Any Node / hook can be a python function that will acquire the GIL

The choice was made here to raise an error as one of the advantage of using cpp extensions with python is to be able to release the GIL. So we prefer to educate users to do it rather than doing it under the hook.

Differential Revision: [D19430979](https://our.internmc.facebook.com/intern/diff/D19430979)

[ghstack-poisoned]

albanD commented:
@zou3519  added the test, can you double check that you're happy with it please?

zou3519 reviewed (approved):
nice!
